### ISSUE TITLE
[Placeholder]: Implement interactive user credential login in Az

### ISSUE THREAD

**Issue by markcowl** [Azure/azure-powershell#7358]
### Description

Work with cli team to design interactive experience for user login with device auth with fallback to current device auth browser experience

- [ ] Write up / prototype design of new functionality
- [ ] Fil;e issues for implementation

### Cost: 8

**Comment by chrisrpatterson** [comment]
It would also be great if the cli and powershell could share the login context so I only need to log one in and I could use both.

**Comment by deathly809** [comment]
@chrisrpatterson Sounds great, but I would be worried about security implications.  It would need to be done very carefully.

**Comment by cormacpayne** [comment]
### General Steps

- Create a new AD application with reply URI pointing at localhost
- Scan for open port (between 8400 and 9000)
- Start a web server that listens for http://localhost:{{PORT}}
- Launch browser that points at the authentication URI
    - If we are unable to launch browser (text browser, no browser, etc.), default back to device code authentication
- Once the user is authenticated, get the authorization code
- Use the authorization code to get the token
- Send the user text in the browser to notify them about authentication result
- Shutdown the web server

### Azure CLI Steps

- [Register the `login` command that is mapped to `login` function](https://github.com/Azure/azure-cli/blob/a52e250279292a0c074c7792fe2a6c1a59881b67/src/command_modules/azure-cli-profile/azure/cli/command_modules/profile/__init__.py#L25)
- [Register the `login` command parameters](https://github.com/Azure/azure-cli/blob/a52e250279292a0c074c7792fe2a6c1a59881b67/src/command_modules/azure-cli-profile/azure/cli/command_modules/profile/__init__.py#L43-L54)
- [`login` function implementation](https://github.com/Azure/azure-cli/blob/a52e250279292a0c074c7792fe2a6c1a59881b67/src/command_modules/azure-cli-profile/azure/cli/command_modules/profile/custom.py#L88-L149)
- [Getting subscriptions in the `login` function by calling the `find_subscription_on_login` function](https://github.com/Azure/azure-cli/blob/a52e250279292a0c074c7792fe2a6c1a59881b67/src/command_modules/azure-cli-profile/azure/cli/command_modules/profile/custom.py#L125-L133)
- [`find_subscription_on_login` function implementation](https://github.com/Azure/azure-cli/blob/32dfb3a37a5a2e8ffd31e465397da08e3a702d5e/src/azure-cli-core/azure/cli/core/_profile.py#L158-L226)
- [Getting subscriptions when device code login is not presented by calling the `find_through_authorization_code_flow` function](https://github.com/Azure/azure-cli/blob/32dfb3a37a5a2e8ffd31e465397da08e3a702d5e/src/azure-cli-core/azure/cli/core/_profile.py#L183-L184)
- [`find_through_authorization_code_flow` function implementation](https://github.com/Azure/azure-cli/blob/32dfb3a37a5a2e8ffd31e465397da08e3a702d5e/src/azure-cli-core/azure/cli/core/_profile.py#L724-L742)
- [Launching the browser and getting the code by calling the `_get_authorization_code` function](https://github.com/Azure/azure-cli/blob/32dfb3a37a5a2e8ffd31e465397da08e3a702d5e/src/azure-cli-core/azure/cli/core/_profile.py#L727)
- [`_get_authorization_code` function implementation](https://github.com/Azure/azure-cli/blob/32dfb3a37a5a2e8ffd31e465397da08e3a702d5e/src/azure-cli-core/azure/cli/core/_profile.py#L1071-L1085)
- [Call to the `_get_authorization_code_worker` function](https://github.com/Azure/azure-cli/blob/32dfb3a37a5a2e8ffd31e465397da08e3a702d5e/src/azure-cli-core/azure/cli/core/_profile.py#L1075-L1076)
- [`_get_authorization_code_worker` function implementation](https://github.com/Azure/azure-cli/blob/32dfb3a37a5a2e8ffd31e465397da08e3a702d5e/src/azure-cli-core/azure/cli/core/_profile.py#L1020-L1068)
- [Find an open port for the `reply_url` between 8400 and 9000](https://github.com/Azure/azure-cli/blob/32dfb3a37a5a2e8ffd31e465397da08e3a702d5e/src/azure-cli-core/azure/cli/core/_profile.py#L1023-L1029)
- [Create a web server at http://localhost:{{PORT}}](https://github.com/Azure/azure-cli/blob/32dfb3a37a5a2e8ffd31e465397da08e3a702d5e/src/azure-cli-core/azure/cli/core/_profile.py#L1025)
- [Construct the browser url](https://github.com/Azure/azure-cli/blob/32dfb3a37a5a2e8ffd31e465397da08e3a702d5e/src/azure-cli-core/azure/cli/core/_profile.py#L1036-L1039)
- [Open the browser by calling the `open_page_in_browser` function](https://github.com/Azure/azure-cli/blob/32dfb3a37a5a2e8ffd31e465397da08e3a702d5e/src/azure-cli-core/azure/cli/core/_profile.py#L1040)
- [`open_page_in_browser` function implementation](https://github.com/Azure/azure-cli/blob/a1f8c7e1f463ad84b719caab26e2651130536018/src/azure-cli-core/azure/cli/core/util.py#L275-L294)
- [Once the code has been received, get the token for authentication by calling the `acquire_token_with_authorization_code` function](https://github.com/Azure/azure-cli/blob/a52e250279292a0c074c7792fe2a6c1a59881b67/src/azure-cli-core/azure/cli/core/_profile.py#L742-L743)

**Comment by afaruqi** [comment]
So, if i understand this correctly, if i create an AAD App/Service Principal, I should be able to use this command? 
$cred = Get-credentials
Connect-AzAccount - Credential $cred -TenantID <tenantid> -ServicePrincipal

Where do I put the ServicePrincipal ID? How will the cmdlet know which servicePrincipal it needs to use? I get an error when I use the above two commands saying my user id (that i typed in the first line) does not have a serviceprincipal associated to it which is correct since i created the serviceprincipal for an app NOT my user id? How can i make this work?

**Comment by afaruqi** [comment]
Thank you for the response! I've "Imported" the Az module, however, when I go to run the Get-AzADServicePrincipal cmdlet, I get an error saying it's not a recognized name of a cmdlet. any thoughts?

**Comment by cormacpayne** [comment]
@afaruqi would you mind running `Get-Module` to see if the `Az.Resources` module is loaded in your current session? If not, would you mind running `Get-Module -ListAvailable -Name Az*` to see if you have the `Az.Resources` module installed?

**Comment by afaruqi** [comment]
Yes i do. So here's my situation. I'm working on two different environments as I don't want the conversion to effect our current scripts using the AzureRm modules. I created the app/serviceprincipal on the computer that has the AzureRm module installed. I am in a catch-22 situation because when i go to the second computer to retrieve the serviceprincipal, it doesn't return anything because I haven't logged in on that computer yet. The reason I haven't logged in is because I'm trying to use Connect-AzAccount but I can't use U/P combo with that anymore. The lack of the backward compatibility with the Connect-AzAccount is really making us go around in circles.

**Comment by cormacpayne** [comment]
@afaruqi correct, U/P is not _currently_ supported in the `Az` module ([more information](https://github.com/Azure/azure-powershell/issues/7430#issuecomment-426480499)), but is there a reason you aren't able to use the (default) device code flow authentication that `Connect-AzAccount` provides? This will tell you to visit a site (more than likely https://microsoft.com/devicelogin, based on your environment), enter the provided code, and then you can sign in to your account from there. Once authenticated, you should be able to perform the above operations to create a service principal in the tenant that your default context is in (this can be found by running `(Get-AzContext).Tenant.Id`).

**Comment by afaruqi** [comment]
Using the device code is NOT feasible for us. We have an automated script that runs to create resources in azure and upload our data to storage, Having to visit the website and type in the code everytime we (at least 5 different people on 5 different machines) run our scripts is very cumbersome and inefficient. We also do not want to use Certificate authentication as it would require us to install the private key of the cert on each individual's machine who is running these scripts. And we do not want to get into the practice of using a secret to authenticate as it is not a very safe approach either. 

I did, however, try your approach of doing _Connect-AzAccount_ and then trying to get the ServicePrincipal. It works up to the point where I use the PSCredential cmdlet. When i run

$Credential = New-Object -TypeName System.Management.Automation.PSCredential -ArgumentList ( $ServicePrincipal.ApplicationId, {{sp_secret}} )

I get the error "New-Object: Cannot find an overload for "PSCredential" and the argument count:2"

Are there any talks of incorporating U/P back into this process?

**Comment by cormacpayne** [comment]
@afaruqi I believe the issue is that the secret you're passing into the `PSCredential` is a `string` and not a `SecureString`. You can convert this to a `SecureString` by running `ConvertTo-SecureString -String {{sp_secret}} -AsPlainText -Force`.

As for U/P support, we are talking internally in our team, as well as with the authentication team (who owns the underlying library where U/P support is found) about ways to expose this scenario in the `Az` module. I believe we should have a timeline available for the public when we GA the `Az` module on December 18th.

**Comment by athipp** [comment]
As per this - https://github.com/Azure/azure-powershell/blob/master/documentation/announcing-az-module.md#authentication-changes
It would be great if you can include a link or provide context in the above location as to why accepting PSCredential is discouraged. thanks.

**Comment by MassimoSporchia** [comment]
Hi, any news?
Did the milestone got shifted to end of Feb?
I've an automated workflow I'd like to do using the Az Powershell module (I need to update a tag value of a resource I just created, which is impossible through the Azure CLI) and this is blocking.

**Comment by afaruqi** [comment]
any update on this issue? Last we talked you were going to fix the Connect-AzAccount cmdlet so it takes in the user credentials. Has that been done?

**Comment by cormacpayne** [comment]
@afaruqi the `-Credentials` parameter is now supported in the `Az` module for _desktop only_ ([more information here](https://github.com/Azure/azure-powershell/issues/7430#issuecomment-426480499))

**Comment by cormacpayne** [comment]
We have an initial prototype with this interactive login functionality working on PowerShell Core, so this will be released when the `Az` module fully migrates to the `Microsoft.Identity.Client` authentication library (MSAL), which we expect to happen in within the next few months. Any additional updates will be posted in this issue.

**Comment by cormacpayne** [comment]
A preview for this functionality can now be found in [2.0.0-preview of `Az.Accounts`](https://www.powershellgallery.com/packages/Az.Accounts/2.0.0-preview). Please see the following [GitHub release](https://github.com/Azure/azure-powershell/releases/tag/Az.Accounts-v2.0.0-preview) for more information on this preview version of the module.

**Comment by markcowl** [comment]
reopening to track GA release

**Comment by TheSaltyKorean** [comment]
Does this solve the issue of the login window with lost focus?

**Comment by chrisedebo** [comment]
I just came across this issue after searching for Connect-Azaccount VSCode having run in to a "hung" session when I executed connect-azaccount. Although I arrived here via two other issues: 

* https://github.com/Azure/azure-powershell/issues/8476 
* https://github.com/Azure/azure-powershell/issues/8649

Both of which describe the focus lost login window. I'm using powershell 5.1.18632 and vscode 1.45.1

Is there a fix for the lost focus login window?
